onnxruntime/core/providers/cuda/math/matmul.cc

```diff
@@ -33,40 +33,116 @@ REGISTER_KERNEL_TYPED(float)
+      return false;
+    }
+  }
+
+  int64_t right_k = transb ? right_shape[right_num_dims - 1] : right_shape[right_num_dims - 2];
+  if (left_k != right_k) {
+    return false;
+  }
+
+  int64_t n = transa ? left_shape[left_num_dims - 1] : left_shape[left_num_dims - 2];
+  int64_t m = transb ? right_shape[right_num_dims - 2] : right_shape[right_num_dims - 1];
+  stride_A = n * left_k;
+  stride_B = right_num_dims == 2 ? 0 : right_k * m;
+  stride_C = n * m;
+  batch_count = left_p;
+  return true;
+}
+
 template <typename T>
 Status MatMul<T>::ComputeInternal(OpKernelContext* ctx) const {
   typedef typename ToCudaType<T>::MappedType CudaT;
 
   const Tensor* left_X = ctx->Input<Tensor>(0);
   const Tensor* right_X = ctx->Input<Tensor>(1);
 
+  // Ignore the transpose flag if rank of input being 1.
+  // Be noted: numpy.transpose on vector does not change anything.
+  bool transa = trans_A_;
+  bool transb = trans_B_;
+  if (left_X->Shape().NumDimensions() == 1) {
+    transa = false;
+  }
+  if (right_X->Shape().NumDimensions() == 1) {
+    transb = false;
+  }
+
   MatMulComputeHelper helper;
-  ORT_RETURN_IF_ERROR(helper.Compute(left_X->Shape(), right_X->Shape()));
+  ORT_RETURN_IF_ERROR(helper.Compute(left_X->Shape(), right_X->Shape(), transa, transb));
 
   Tensor* Y = ctx->Output(0, helper.OutputShape());
   ORT_RETURN_IF_NOT(strcmp(Y->Location().name, CUDA) == 0, "Output should be allocated on CUDA");
 
   CudaT one = ToCudaType<T>::FromFloat(1.0f);
   CudaT zero = ToCudaType<T>::FromFloat(0.0f);
 
+  cublasOperation_t transA = transa ? CUBLAS_OP_T : CUBLAS_OP_N;
+  cublasOperation_t transB = transb ? CUBLAS_OP_T : CUBLAS_OP_N;
+  const int lda = transa ? static_cast<int>(helper.M()) : static_cast<int>(helper.K());
+  const int ldb = transb ? static_cast<int>(helper.K()) : static_cast<int>(helper.N());
+  const int ldc = static_cast<int>(helper.N());
+  int64_t stride_A, stride_B, stride_C, batch_count;
+
   if (helper.OutputOffsets().size() == 1) {
     CUBLAS_RETURN_IF_ERROR(cublasGemmHelper(
         Base::CublasHandle(),
-        CUBLAS_OP_N,
-        CUBLAS_OP_N,
+        transB,
+        transA,
         static_cast<int>(helper.N()),
         static_cast<int>(helper.M()),
         static_cast<int>(helper.K()),
         &one,
         reinterpret_cast<const CudaT*>(right_X->template Data<T>()),
-        static_cast<int>(helper.N()),
+        ldb,
         reinterpret_cast<const CudaT*>(left_X->template Data<T>()),
-        static_cast<int>(helper.K()),
+        lda,
         &zero,
         reinterpret_cast<CudaT*>(Y->template MutableData<T>()),
-        static_cast<int>(helper.N())));
+        ldc));
+    return Status::OK();
+  } else if (CanUseStridedBatchedGemm(left_X->Shape(), right_X->Shape(),
+             transa, transb, stride_A, stride_B, stride_C, batch_count)) {
+    CUBLAS_RETURN_IF_ERROR(cublasGemmStridedBatchedHelper(Base::CublasHandle(),
+                                                          transB,
```

Comment:
Are there unit tests that hit this new code?

Reply:
will revert it for now, in the next merge we'll include it with proper structure